### PR TITLE
Cache varstore

### PR DESCRIPTION
**Pull request by behdad**
Fixes issue originally raised in https://github.com/harfbuzz/harfbuzz/pull/2878#issuecomment-785340348

Hi Jonathan,

Can you test this in Firefox and see if it helps with variable font performance?

Thanks

### CODE REVIEW

**Comment by behdad** [comment]
I see 4% speedup in shaping of a simple variable font, and 32% speedup in shaping of RobotoFlex, which is a heavy variable font.

Non-variable font shaping performance is unaffected.

These numbers are with hb font-funcs.

Just looking at advance-width speedup, for simple variable font, 30% speedup, for RobotoFlex, 50%.

**Comment by drott** [comment]
Thanks for doing this optimisation.

**Review by drott — COMMENTED** [review]
🤦‍♂️ Github ate my homework, I meant to send this review with minor comments. LGTM with some remarks. Anyway, thanks again!

File: src/hb-ot-hmtx-table.hh [hunk 1]
```diff
@@ -310,7 +311,7 @@ struct hmtxvmtx
 
     unsigned int default_advance;
 
-    private:
+    public:
```
[the remark this answers is not in the document]
Comment (behdad) [reply src/hb-ot-hmtx-table.hh:314]: `var_table` is now accessed from `hb-ot-font` to access the VarStore to create the cache.

File: src/hb-ot-layout-common.hh [hunk 2]
```diff
@@ -2601,14 +2601,27 @@ struct VarRegionAxis
...
+
   float evaluate (unsigned int region_index,
-		  const int *coords, unsigned int coord_len) const
+		  const int *coords, unsigned int coord_len,
+		  cache_t *cache = nullptr) const
   {
     if (unlikely (region_index >= regionCount))
       return 0.;
 
+    float *cached = nullptr;
```
[the remark this answers is not in the document]
Comment (behdad) [reply src/hb-ot-layout-common.hh:2617]: Good point. Will do.

File: src/hb-ot-var-hvar-table.hh [hunk 3]
```diff
@@ -335,7 +340,7 @@ struct HVARVVAR
 
   bool has_side_bearing_deltas () const { return lsbMap && rsbMap; }
 
-  protected:
+  public:
```
[the remark this answers is not in the document]
Comment (behdad) [reply src/hb-ot-var-hvar-table.hh:343]: Same. Need to reach out to the varStore. Can add accessors. But these are safe to access. In those cases we just make them public.

**Comment by jfkthame** [comment]
Hi @behdad - sorry to be so late responding to this, I've been out for a while (first PTO and then Covid)... trying to catch up on things. I tried a Firefox build with this applied, and can confirm that it appears to help. For a huge reflow (a 40MB text block) using the Apple system font, which is variable (but not very complex), it reduced the time spent under VariationDevice during GPOS pair-pos processing from around 500ms to 300ms. Obviously the specific gains will be depend on characteristics of the font being used, but in any case it's a nice win - thanks!

**Comment by behdad** [comment]
Thanks @jfkthame

**Pull request merged**